### PR TITLE
add a faster pkg_tar builder

### PR DESCRIPTION
**Pull request by mikedanese**
This implements the parts of pkg_tar builder that we use and is way faster. I can build all of release-tars on my workstation in ~35 seconds.

### CODE REVIEW

**Comment by BenTheElder** [comment]
Aww yes!
We need something like the opposite of shrug @krzyzacy.

File: tools/fast_build_tar/buildtar.go [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+	uname := ""
+	gname := ""
+
+	info, err := os.Stat(file)
+	if err != nil {
+		return err
+	}
+	header := &tar.Header{
+		Name:       dest,
+		Mode:       int64(mode),
```
Comment (rmmh) [inline tools/fast_build_tar/buildtar.go]: TODO: fix mode based on info?

**Comment by mikedanese** [comment]
This passes all the pkg_tar tests excluding bz2 and xz compression.

**Comment by mikedanese** [comment]
These are the kubernetes changes needed.

https://github.com/kubernetes/kubernetes/compare/master...mikedanese:fastpack?expand=1

File: tools/fast_build_tar/buildtar.go (hunk 1 above)
Comment (mikedanese) [reply tools/fast_build_tar/buildtar.go]: This matches the bazel/build_tar.py behavior. I'm trying to keep these compatible, at least initially.

File: tools/fast_build_tar/buildtar.go [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+
+	header := tar.Header{
+		Name:       dest,
+		Mode:       int64(mode),
+		Uid:        uid,
+		Gid:        gid,
+		Size:       0,
+		Uname:      uname,
+		Gname:      gname,
+		AccessTime: time.Now(),
```
Comment (ixdy) [inline tools/fast_build_tar/buildtar.go]: I'm pretty sure the reference version always uses a timestamp of 0 (for reproducibility).

**Comment by ixdy** [comment]
also something is acting weirdly; if I look at one of the tarfiles generated by this tool, at the end `tar` prints
```
tar: Unexpected EOF in archive
tar: Error is not recoverable: exiting now
```

**Comment by mikedanese** [comment]
@ixdy fixed.

**Comment by BenTheElder** [comment]
@mikedanese looks like it failed the linter in Travis:
```
tools/fast_build_tar/buildtar.go:74:26:warning: error strings should not be capitalized or end with punctuation or a newline (golint)
tools/fast_build_tar/buildtar.go:203:21:warning: error strings should not be capitalized or end with punctuation or a newline (golint)
tools/fast_build_tar/buildtar.go:476:14:warning: struct field defaultUid should be defaultUID (golint)
tools/fast_build_tar/buildtar.go:493:20:warning: method getUid should be getUID (golint)
tools/fast_build_tar/buildtar.go:221::warning: declaration of "err" shadows declaration at tools/fast_build_tar/buildtar.go:210 (vetshadow)
```
¯\\\_(ツ)\_/¯

File: tools/fast_build_tar/buildtar.go [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+		gzr, err := gzip.NewReader(r)
+		if err != nil {
+			return err
+		}
+		r = gzr
+	case strings.HasSuffix(toAdd, "bz2"):
+		bz2r := bzip2.NewReader(r)
+		r = bz2r
+	case strings.HasSuffix(toAdd, "xz"):
+		return fmt.Errorf("%q decompression is not supported yet. File a bug to get it implemented!", toAdd)
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply tools/fast_build_tar/buildtar.go]: this is bs

File: tools/fast_build_tar/buildtar.go [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+	w = buf
+
+	switch compression {
+	case "":
+	case "gz":
+		gzw := pargzip.NewWriter(w)
+		closers = append(closers, func() { gzw.Close() })
+		w = gzw
+	case "bz2", "xz":
+		return nil, fmt.Errorf("%q compression is not supported yet. File a bug to get it implemented!", compression)
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply tools/fast_build_tar/buildtar.go]: this is bs

**Comment by mikedanese** [comment]
@ixdy ping

File: tools/fast_build_tar/buildtar.go [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+
+	"golang.org/x/build/pargzip"
+)
+
+type tarFile struct {
+	directory string
```
Comment (ixdy) [inline tools/fast_build_tar/buildtar.go]: what is this used for?

File: tools/fast_build_tar/buildtar.go [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+
+	flag.Parse()
+
+	if flagfile != "" {
+		b, err := ioutil.ReadFile(flagfile)
+		if err != nil {
+			glog.Fatalf("couldn't read flagfile: %v", err)
+		}
+		cmdline := strings.Split(string(b), "\n")
+		flag.CommandLine.Parse(cmdline)
```
Comment (ixdy) [inline tools/fast_build_tar/buildtar.go]: does this mean that the flagfile overrides the cmdline args? is that what we want? (maybe we don't care?)

**Comment by ixdy** [comment]
cc @damienmg @mattmoor

File: tools/fast_build_tar/buildtar.go (hunk 5 above)
Comment (mikedanese) [reply tools/fast_build_tar/buildtar.go]: it's the directory in which to store the file inside the layer

File: tools/fast_build_tar/buildtar.go (hunk 6 above)
Comment (mikedanese) [reply tools/fast_build_tar/buildtar.go]: It overrides non multistring args, yes. These is very close to python flagfile but I didn't want to rewrite the parser. It's not an exact match. gflag says:

> Flags are always processed in the expected order. That is, processing begins by examining the flags specified directly on the command line. If a flagfile is specified, its contents are processed, and then processing continues with remaining flags from the command line.

**Pull request merged**